File: java/refactoring-methods/04-data-organization.java
import java.util.*;

/**
 * 18. Self-Encapsulate Field
 *
 * BEFORE: Direct field access
 */
class PersonBefore {
    public String name; // Direct access

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}

/**
 * AFTER: Self-encapsulate field
 */
class PersonAfter {
    private String name;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}

/**
 * 19. Replacing the data value with an object (Replace Data Value with Object)
 *
 * BEFORE: Primitive data type that should be an object
 */
class OrderBefore {
    private String customer; // Just a string

    public String getCustomerName() {
        return customer;
    }

    public void setCustomer(String customer) {
        this.customer = customer;
    }
}

/**
 * AFTER: Replace with object
 */
class Customer {
    private String name;

    public Customer(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }
}

class OrderAfter {
    private Customer customer;

    public Customer getCustomer() {
        return customer;
    }

    public void setCustomer(Customer customer) {
        this.customer = customer;
    }

    public String getCustomerName() {
        return customer.getName();
    }
}

/**
 * 20. Replacing the value with a reference (Change Value to Reference)
 *
 * BEFORE: Multiple instances of same object
 */
class CustomerValue {
    private String name;

    public CustomerValue(String name) {
        this.name = name;
    }

    public String getName() {
        return name;
    }
}

class OrderValue {
    private CustomerValue customer; // New instance for each order

    public OrderValue(String customerName) {
        this.customer = new CustomerValue(customerName);
    }
}

/**
 * AFTER: Use reference to single instance
 */
class CustomerReference {
    private String name;
    private static Map<String, CustomerReference> instances = new HashMap<>();

    private CustomerReference(String name) {
        this.name = name;
    }

    public static CustomerReference create(String name) {
        if (!instances.containsKey(name)) {
            instances.put(name, new CustomerReference(name));
        }
        return instances.get(name);
    }

    public String getName() {
        return name;
    }
}

class OrderReference {
    private CustomerReference customer;

    public OrderReference(String customerName) {
        this.customer = CustomerReference.create(customerName);
    }
}

/**
 * 21. Replacing a reference with a value (Change Reference to Value)
 *
 * BEFORE: Unnecessary reference when value would suffice
 */
class CurrencyReference {
    private String code;

    public CurrencyReference(String code) {
        this.code = code;
    }

    public String getCode() {
        return code;
    }
}

class ProductReference {
    private double price;
    private CurrencyReference currency; // Reference object

    public ProductReference(double price, CurrencyReference currency) {
        this.price = price;
        this.currency = currency;
    }
}

/**
 * AFTER: Use value object instead
 */
class CurrencyValue {
    private String code;

    public CurrencyValue(String code) {
        this.code = code;
    }

    public String getCode() {
        return code;
    }
}

class ProductValue {
    private double price;
    private String currencyCode; // Just the value

    public ProductValue(double price, String currencyCode) {
        this.price = price;
        this.currencyCode = currencyCode;
    }

    public String getCurrencyCode() {
        return currencyCode;
    }
}

/**
 * 22. Replacing an array with an object (Replace Array with Object)
 *
 * BEFORE: Using array for structured data
 */
class PerformanceArray {
    public Map<String, Object> getPerformanceData() {
        Map<String, Object> data = new HashMap<>();
        data.put("goals", 10);
        data.put("assists", 5);
        data.put("minutes", 120);
        return data;
    }

    public double calculateScore(Map<String, Object> data) {
        int goals = (Integer) data.get("goals");
        int assists = (Integer) data.get("assists");
        int minutes = (Integer) data.get("minutes");
        return (goals * 2) + (assists * 1.5) + (minutes / 60.0);
    }
}

/**
 * AFTER: Replace array with object
 */
class PerformanceData {
    private int goals;
    private int assists;
    private int minutes;

    public PerformanceData(int goals, int assists, int minutes) {
        this.goals = goals;
        this.assists = assists;
        this.minutes = minutes;
    }

    public int getGoals() {
        return goals;
    }

    public int getAssists() {
        return assists;
    }

    public int getMinutes() {
        return minutes;
    }

    public double calculateScore() {
        return (goals * 2) + (assists * 1.5) + (minutes / 60.0);
    }
}

class PerformanceObject {
    public PerformanceData getPerformanceData() {
        return new PerformanceData(10, 5, 120);
    }

    public double calculateScore(PerformanceData data) {
        return data.calculateScore();
    }
}

/**
 * 23. Duplication of visible data (Duplicate Observed Data)
 *
 * BEFORE: Domain data mixed with presentation
 */
class OrderDomain {
    private double total = 0;

    public void addItem(double price) {
        total += price;
        // Have to update UI here too
        updateDisplay();
    }

    private void updateDisplay() {
        // Update UI elements
    }
}

/**
 * AFTER: Separate domain and presentation data
 */
interface OrderObserver {
    void update(double total);
}

class OrderDomainSeparated {
    private double total = 0;
    private List<OrderObserver> observers = new ArrayList<>();

    public void addItem(double price) {
        total += price;
        notifyObservers();
    }

    public double getTotal() {
        return total;
    }

    public void addObserver(OrderObserver observer) {
        observers.add(observer);
    }

    private void notifyObservers() {
        for (OrderObserver observer : observers) {
            observer.update(total);
        }
    }
}

class OrderDisplay implements OrderObserver {
    private OrderDomainSeparated order;

    public OrderDisplay(OrderDomainSeparated order) {
        this.order = order;
        order.addObserver(this);
    }

    public void update(double total) {
        // Update display with new total
    }
}

/**
 * 24. Replacing Unidirectional communication with Bidirectional
 * communication (Change Unidirectional Association to Bidirectional)
 *
 * BEFORE: One-way association
 */
class CustomerUni {
    private List<OrderUni> orders = new ArrayList<>();

    public void addOrder(OrderUni order) {
        orders.add(order);
        // Order doesn't know about customer
    }
}

class OrderUni {
    private List<String> items = new ArrayList<>();
}

/**
 * AFTER: Bidirectional association
 */
class CustomerBi {
    private List<OrderBi> orders = new ArrayList<>();

    public void addOrder(OrderBi order) {
        orders.add(order);
        order.setCustomer(this);
    }
}

class OrderBi {
    private CustomerBi customer;
    private List<String> items = new ArrayList<>();

    public void setCustomer(CustomerBi customer) {
        this.customer = customer;
    }

    public CustomerBi getCustomer() {
        return customer;
    }
}

/**
 * 25. Replacing Bidirectional communication with Unidirectional
 * communication (Change Bidirectional Association to Unidirectional)
 *
 * BEFORE: Unnecessary bidirectional association
 */
class CustomerBidirectional {
    private List<OrderBidirectional> orders = new ArrayList<>();

    public void addOrder(OrderBidirectional order) {
        orders.add(order);
        order.setCustomer(this);
    }
}

class OrderBidirectional {
    private CustomerBidirectional customer;

    public void setCustomer(CustomerBidirectional customer) {
        this.customer = customer;
    }

    public CustomerBidirectional getCustomer() {
        return customer;
    }
}

/**
 * AFTER: Remove bidirectional link
 */
class CustomerUnidirectional {
    private List<OrderUnidirectional> orders = new ArrayList<>();

    public void addOrder(OrderUnidirectional order) {
        orders.add(order);
    }
}

class OrderUnidirectional {
    private String customerId;

    public OrderUnidirectional(String customerId) {
        this.customerId = customerId;
    }

    public String getCustomerId() {
        return customerId;
    }
}

/**
 * 26. Replacing the magic number with a symbolic constant
 * (Replace Magic Number with Symbolic Constant)
 *
 * BEFORE: Magic numbers
 */
class GeometryBefore {
    public double calculateCircleArea(double radius) {
        return 3.14159 * radius * radius; // Magic number
    }

    public double calculateCircleCircumference(double radius) {
        return 2 * 3.14159 * radius; // Same magic number
    }
}

/**
 * AFTER: Use symbolic constant
 */
class GeometryAfter {
    public static final double PI = 3.14159;

    public double calculateCircleArea(double radius) {
        return PI * radius * radius;
    }

    public double calculateCircleCircumference(double radius) {
        return 2 * PI * radius;
    }
}

/**
 * 27. Encapsulate Field
 *
 * BEFORE: Public field
 */
class PersonPublic {
    public String name;
}

/**
 * AFTER: Encapsulated field
 */
class PersonEncapsulated {
    private String name;

    public String getName() {
        return name;
    }

    public void setName(String name) {
        this.name = name;
    }
}

/**
 * 28. Encapsulate Collection
 *
 * BEFORE: Direct access to collection
 */
class TeamBefore {
    public List<String> players = new ArrayList<>(); // Direct access

    public void addPlayer(String player) {
        players.add(player);
    }
}

/**
 * AFTER: Encapsulated collection
 */
class TeamAfter {
    private List<String> players = new ArrayList<>();

    public void addPlayer(String player) {
        players.add(player);
    }

    public void removePlayer(String player) {
        players.remove(player);
    }

    public List<String> getPlayers() {
        return new ArrayList<>(players); // Return copy
    }

    public int getPlayerCount() {
        return players.size();
    }
}

/**
 * 29. Replacing a record with a Data Class
 *
 * BEFORE: Using array as data structure
 */
class EmployeeArray {
    public Map<String, Object> createEmployee(Map<String, Object> data) {
        Map<String, Object> employee = new HashMap<>();
        employee.put("name", data.get("name"));
        employee.put("salary", data.get("salary"));
        employee.put("department", data.get("department"));
        return employee;
    }

    public double getSalary(Map<String, Object> employee) {
        return (Double) employee.get("salary");
    }
}

/**
 * AFTER: Use data class
 */
class Employee {
    private String name;
    private double salary;
    private String department;

    public Employee(String name, double salary, String department) {
        this.name = name;
        this.salary = salary;
        this.department = department;
    }

    public String getName() {
        return name;
    }

    public double getSalary() {
        return salary;
    }

    public String getDepartment() {
        return department;
    }
}

class EmployeeDataClass {
    public Employee createEmployee(String name, double salary, String department) {
        return new Employee(name, salary, department);
    }

    public double getSalary(Employee employee) {
        return employee.getSalary();
    }
}

/**
 * 30. Replacing Type Code with Class
 *
 * BEFORE: Type code as constants
 */
class EmployeeTypeCode {
    public static final int ENGINEER = 0;
    public static final int SALESMAN = 1;
    public static final int MANAGER = 2;

    private int type;

    public EmployeeTypeCode(int type) {
        this.type = type;
    }

    public int getTypeCode() {
        return type;
    }

    public double getMonthlySalary() {
        switch (type) {
            case ENGINEER:
                return 5000;
            case SALESMAN:
                return 4000;
            case MANAGER:
                return 6000;
            default:
                return 0;
        }
    }
}

/**
 * AFTER: Replace type code with class
 */
abstract class EmployeeType {
    public abstract double getMonthlySalary();

    public static EngineerType createEngineer() {
        return new EngineerType();
    }

    public static SalesmanType createSalesman() {
        return new SalesmanType();
    }

    public static ManagerType createManager() {
        return new ManagerType();
    }
}

class EngineerType extends EmployeeType {
    public double getMonthlySalary() {
        return 5000;
    }
}

class SalesmanType extends EmployeeType {
    public double getMonthlySalary() {
        return 4000;
    }
}

class ManagerType extends EmployeeType {
    public double getMonthlySalary() {
        return 6000;
    }
}

class EmployeeTypeClass {
    private EmployeeType type;

    public EmployeeTypeClass(EmployeeType type) {
        this.type = type;
    }

    public double getMonthlySalary() {
        return type.getMonthlySalary();
    }
}

/**
 * 31. Replacing Type Code with Subclasses
 *
 * BEFORE: Type code in base class
 */
class EmployeeSubBefore {
    public static final int ENGINEER = 0;
    public static final int SALESMAN = 1;
    public static final int MANAGER = 2;

    private int type;
    private double salary;

    public EmployeeSubBefore(int type, double salary) {
        this.type = type;
        this.salary = salary;
    }

    public double getSalary() {
        return salary;
    }

    public int getType() {
        return type;
    }
}

/**
 * AFTER: Replace type code with subclasses
 */
abstract class EmployeeSubAfter {
    protected double salary;

    public EmployeeSubAfter(double salary) {
        this.salary = salary;
    }

    public double getSalary() {
        return salary;
    }

    public abstract String getType();
}

class Engineer extends EmployeeSubAfter {
    public Engineer(double salary) {
        super(salary);
    }

    public String getType() {
        return "engineer";
    }
}

class Salesman extends EmployeeSubAfter {
    public Salesman(double salary) {
        super(salary);
    }

    public String getType() {
        return "salesman";
    }
}

class Manager extends EmployeeSubAfter {
    public Manager(double salary) {
        super(salary);
    }

    public String getType() {
        return "manager";
    }
}

/**
 * 32. Replacing Type Code with State/Strategy
 *
 * BEFORE: Type code with behavior
 */
class EmployeeStateBefore {
    public static final int JUNIOR = 0;
    public static final int SENIOR = 1;
    public static final int LEAD = 2;

    private int level;

    public EmployeeStateBefore(int level) {
        this.level = level;
    }

    public double getSalaryMultiplier() {
        switch (level) {
            case JUNIOR:
                return 1.0;
            case SENIOR:
                return 1.5;
            case LEAD:
                return 2.0;
            default:
                return 1.0;
        }
    }
}

/**
 * AFTER: Use state/strategy pattern
 */
interface EmployeeLevel {
    double getSalaryMultiplier();
}

class JuniorLevel implements EmployeeLevel {
    public double getSalaryMultiplier() {
        return 1.0;
    }
}

class SeniorLevel implements EmployeeLevel {
    public double getSalaryMultiplier() {
        return 1.5;
    }
}

class LeadLevel implements EmployeeLevel {
    public double getSalaryMultiplier() {
        return 2.0;
    }
}

class EmployeeStateAfter {
    private EmployeeLevel level;

    public EmployeeStateAfter(EmployeeLevel level) {
        this.level = level;
    }

    public double getSalaryMultiplier() {
        return level.getSalaryMultiplier();
    }
}

/**
 * 33. Replacing Subclass with Fields
 *
 * BEFORE: Unnecessary subclasses
 */
abstract class PersonSub {
    protected String name;
    protected String gender;

    public PersonSub(String name, String gender) {
        this.name = name;
        this.gender = gender;
    }

    public String getName() {
        return name;
    }

    public abstract boolean isMale();
}

class Male extends PersonSub {
    public Male(String name) {
        super(name, "male");
    }

    public boolean isMale() {
        return true;
    }
}

class Female extends PersonSub {
    public Female(String name) {
        super(name, "female");
    }

    public boolean isMale() {
        return false;
    }
}

/**
 * AFTER: Replace subclass with field
 */
class PersonField {
    private String name;
    private String gender; // 'male' or 'female'

    public PersonField(String name, String gender) {
        this.name = name;
        this.gender = gender;
    }

    public String getName() {
        return name;
    }

    public boolean isMale() {
        return "male".equals(gender);
    }

    public String getGender() {
        return gender;
    }
}

File: python/refactoring-methods/04-data-organization.py
from abc import ABC, abstractmethod

# 18. Self-Encapsulate Field
#
# BEFORE: Direct field access
class PersonBefore:
    def __init__(self, name):
        self.name = name  # Direct access

    def get_name(self):
        return self.name

    def set_name(self, name):
        self.name = name

# AFTER: Self-encapsulate field
class PersonAfter:
    def __init__(self, name):
        self._name = name

    def get_name(self):
        return self._name

    def set_name(self, name):
        self._name = name

# 19. Replacing the data value with an object (Replace Data Value with Object)
#
# BEFORE: Primitive data type that should be an object
class OrderBefore:
    def __init__(self, customer):
        self._customer = customer  # Just a string

    def get_customer_name(self):
        return self._customer

    def set_customer(self, customer):
        self._customer = customer

# AFTER: Replace with object
class Customer:
    def __init__(self, name):
        self._name = name

    def get_name(self):
        return self._name

class OrderAfter:
    def __init__(self, customer):
        self._customer = customer

    def get_customer(self):
        return self._customer

    def set_customer(self, customer):
        self._customer = customer

    def get_customer_name(self):
        return self._customer.get_name()

# 20. Replacing the value with a reference (Change Value to Reference)
#
# BEFORE: Multiple instances of same object
class CustomerValue:
    def __init__(self, name):
        self._name = name

    def get_name(self):
        return self._name

class OrderValue:
    def __init__(self, customer_name):
        self._customer = CustomerValue(customer_name)  # New instance for each order

# AFTER: Use reference to single instance
class CustomerReference:
    _instances = {}

    def __init__(self, name):
        self._name = name

    @classmethod
    def create(cls, name):
        if name not in cls._instances:
            cls._instances[name] = cls(name)
        return cls._instances[name]

    def get_name(self):
        return self._name

class OrderReference:
    def __init__(self, customer_name):
        self._customer = CustomerReference.create(customer_name)

# 21. Replacing a reference with a value (Change Reference to Value)
#
# BEFORE: Unnecessary reference when value would suffice
class CurrencyReference:
    def __init__(self, code):
        self._code = code

    def get_code(self):
        return self._code

class ProductReference:
    def __init__(self, price, currency):
        self._price = price
        self._currency = currency  # Reference object

# AFTER: Use value object instead
class CurrencyValue:
    def __init__(self, code):
        self._code = code

    def get_code(self):
        return self._code

class ProductValue:
    def __init__(self, price, currency_code):
        self._price = price
        self._currency_code = currency_code  # Just the value

    def get_currency_code(self):
        return self._currency_code

# 22. Replacing an array with an object (Replace Array with Object)
#
# BEFORE: Using array for structured data
class PerformanceArray:
    def get_performance_data(self):
        return {
            'goals': 10,
            'assists': 5,
            'minutes': 120
        }

    def calculate_score(self, data):
        return (data['goals'] * 2) + (data['assists'] * 1.5) + (data['minutes'] / 60)

# AFTER: Replace array with object
class PerformanceData:
    def __init__(self, goals, assists, minutes):
        self._goals = goals
        self._assists = assists
        self._minutes = minutes

    def get_goals(self):
        return self._goals

    def get_assists(self):
        return self._assists

    def get_minutes(self):
        return self._minutes

    def calculate_score(self):
        return (self._goals * 2) + (self._assists * 1.5) + (self._minutes / 60)

class PerformanceObject:
    def get_performance_data(self):
        return PerformanceData(10, 5, 120)

    def calculate_score(self, data):
        return data.calculate_score()

# 23. Duplication of visible data (Duplicate Observed Data)
#
# BEFORE: Domain data mixed with presentation
class OrderDomain:
    def __init__(self):
        self._total = 0

    def add_item(self, price):
        self._total += price
        # Have to update UI here too
        self._update_display()

    def _update_display(self):
        # Update UI elements
        pass

# AFTER: Separate domain and presentation data
class OrderDomainSeparated:
    def __init__(self):
        self._total = 0
        self._observers = []

    def add_item(self, price):
        self._total += price
        self._notify_observers()

    def get_total(self):
        return self._total

    def add_observer(self, observer):
        self._observers.append(observer)

    def _notify_observers(self):
        for observer in self._observers:
            observer.update(self._total)

class OrderDisplay:
    def __init__(self, order):
        self._order = order
        self._order.add_observer(self)

    def update(self, total):
        # Update display with new total
        pass

# 24. Replacing Unidirectional communication with Bidirectional
# communication (Change Unidirectional Association to Bidirectional)
#
# BEFORE: One-way association
class CustomerUni:
    def __init__(self):
        self._orders = []

    def add_order(self, order):
        self._orders.append(order)
        # Order doesn't know about customer

class OrderUni:
    def __init__(self):
        self._items = []

# AFTER: Bidirectional association
class CustomerBi:
    def __init__(self):
        self._orders = []

    def add_order(self, order):
        self._orders.append(order)
        order.set_customer(self)

class OrderBi:
    def __init__(self):
        self._customer = None
        self._items = []

    def set_customer(self, customer):
        self._customer = customer

    def get_customer(self):
        return self._customer

# 25. Replacing Bidirectional communication with Unidirectional
# communication (Change Bidirectional Association to Unidirectional)
#
# BEFORE: Unnecessary bidirectional association
class CustomerBidirectional:
    def __init__(self):
        self._orders = []

    def add_order(self, order):
        self._orders.append(order)
        order.set_customer(self)

class OrderBidirectional:
    def __init__(self):
        self._customer = None

    def set_customer(self, customer):
        self._customer = customer

    def get_customer(self):
        return self._customer

# AFTER: Remove bidirectional link
class CustomerUnidirectional:
    def __init__(self):
        self._orders = []

    def add_order(self, order):
        self._orders.append(order)

class OrderUnidirectional:
    def __init__(self, customer_id):
        self._customer_id = customer_id

    def get_customer_id(self):
        return self._customer_id

# 26. Replacing the magic number with a symbolic constant
# (Replace Magic Number with Symbolic Constant)
#
# BEFORE: Magic numbers
class GeometryBefore:
    def calculate_circle_area(self, radius):
        return 3.14159 * radius * radius  # Magic number

    def calculate_circle_circumference(self, radius):
        return 2 * 3.14159 * radius  # Same magic number

# AFTER: Use symbolic constant
class GeometryAfter:
    PI = 3.14159

    def calculate_circle_area(self, radius):
        return self.PI * radius * radius

    def calculate_circle_circumference(self, radius):
        return 2 * self.PI * radius

# 27. Encapsulate Field
#
# BEFORE: Public field
class PersonPublic:
    def __init__(self, name):
        self.name = name  # Public field

# AFTER: Encapsulated field
class PersonEncapsulated:
    def __init__(self, name):
        self._name = name

    def get_name(self):
        return self._name

    def set_name(self, name):
        self._name = name

# 28. Encapsulate Collection
#
# BEFORE: Direct access to collection
class TeamBefore:
    def __init__(self):
        self.players = []  # Direct access

    def add_player(self, player):
        self.players.append(player)

# AFTER: Encapsulated collection
class TeamAfter:
    def __init__(self):
        self._players = []

    def add_player(self, player):
        self._players.append(player)

    def remove_player(self, player):
        if player in self._players:
            self._players.remove(player)

    def get_players(self):
        return self._players.copy()  # Return copy

    def get_player_count(self):
        return len(self._players)

# 29. Replacing a record with a Data Class
#
# BEFORE: Using array as data structure
class EmployeeArray:
    def create_employee(self, data):
        return {
            'name': data['name'],
            'salary': data['salary'],
            'department': data['department']
        }

    def get_salary(self, employee):
        return employee['salary']

# AFTER: Use data class
class Employee:
    def __init__(self, name, salary, department):
        self._name = name
        self._salary = salary
        self._department = department

    def get_name(self):
        return self._name

    def get_salary(self):
        return self._salary

    def get_department(self):
        return self._department

class EmployeeDataClass:
    def create_employee(self, name, salary, department):
        return Employee(name, salary, department)

    def get_salary(self, employee):
        return employee.get_salary()

# 30. Replacing Type Code with Class
#
# BEFORE: Type code as constants
class EmployeeTypeCode:
    ENGINEER = 0
    SALESMAN = 1
    MANAGER = 2

    def __init__(self, employee_type):
        self._type = employee_type

    def get_type_code(self):
        return self._type

    def get_monthly_salary(self):
        if self._type == self.ENGINEER:
            return 5000
        elif self._type == self.SALESMAN:
            return 4000
        elif self._type == self.MANAGER:
            return 6000
        else:
            return 0

# AFTER: Replace type code with class
class EmployeeType(ABC):
    @abstractmethod
    def get_monthly_salary(self):
        pass

    @staticmethod
    def create_engineer():
        return EngineerType()

    @staticmethod
    def create_salesman():
        return SalesmanType()

    @staticmethod
    def create_manager():
        return ManagerType()

class EngineerType(EmployeeType):
    def get_monthly_salary(self):
        return 5000

class SalesmanType(EmployeeType):
    def get_monthly_salary(self):
        return 4000

class ManagerType(EmployeeType):
    def get_monthly_salary(self):
        return 6000

class EmployeeTypeClass:
    def __init__(self, employee_type):
        self._type = employee_type

    def get_monthly_salary(self):
        return self._type.get_monthly_salary()

# 31. Replacing Type Code with Subclasses
#
# BEFORE: Type code in base class
class EmployeeSubBefore:
    ENGINEER = 0
    SALESMAN = 1
    MANAGER = 2

    def __init__(self, employee_type, salary):
        self._type = employee_type
        self._salary = salary

    def get_salary(self):
        return self._salary

    def get_type(self):
        return self._type

# AFTER: Replace type code with subclasses
class EmployeeSubAfter(ABC):
    def __init__(self, salary):
        self._salary = salary

    def get_salary(self):
        return self._salary

    @abstractmethod
    def get_type(self):
        pass

class Engineer(EmployeeSubAfter):
    def get_type(self):
        return 'engineer'

class Salesman(EmployeeSubAfter):
    def get_type(self):
        return 'salesman'

class Manager(EmployeeSubAfter):
    def get_type(self):
        return 'manager'

# 32. Replacing Type Code with State/Strategy
#
# BEFORE: Type code with behavior
class EmployeeStateBefore:
    JUNIOR = 0
    SENIOR = 1
    LEAD = 2

    def __init__(self, level):
        self._level = level

    def get_salary_multiplier(self):
        if self._level == self.JUNIOR:
            return 1.0
        elif self._level == self.SENIOR:
            return 1.5
        elif self._level == self.LEAD:
            return 2.0
        else:
            return 1.0

# AFTER: Use state/strategy pattern
class EmployeeLevel(ABC):
    @abstractmethod
    def get_salary_multiplier(self):
        pass

class JuniorLevel(EmployeeLevel):
    def get_salary_multiplier(self):
        return 1.0

class SeniorLevel(EmployeeLevel):
    def get_salary_multiplier(self):
        return 1.5

class LeadLevel(EmployeeLevel):
    def get_salary_multiplier(self):
        return 2.0

class EmployeeStateAfter:
    def __init__(self, level):
        self._level = level

    def get_salary_multiplier(self):
        return self._level.get_salary_multiplier()

# 33. Replacing Subclass with Fields
#
# BEFORE: Unnecessary subclasses
class PersonSub(ABC):
    def __init__(self, name, gender):
        self._name = name
        self._gender = gender

    def get_name(self):
        return self._name

    @abstractmethod
    def is_male(self):
        pass

class Male(PersonSub):
    def is_male(self):
        return True

class Female(PersonSub):
    def is_male(self):
        return False

# AFTER: Replace subclass with field
class PersonField:
    def __init__(self, name, gender):
        self._name = name
        self._gender = gender  # 'male' or 'female'

    def get_name(self):
        return self._name

    def is_male(self):
        return self._gender == 'male'

    def get_gender(self):
        return self._gender
